pkg/tanka/parallel.go

```diff
@@ -4,79 +4,87 @@ import (
 	"fmt"
 	"sync"
 
+	"k8s.io/apimachinery/pkg/labels"
+
 	"github.com/grafana/tanka/pkg/spec/v1alpha1"
 )
 
 const parallel = 8
 
+type parallelOpts struct {
+	JsonnetOpts JsonnetOpts
+	Selector    labels.Selector
+	Parallel    int
+}
+
 // parallelLoadEnvironments evaluates multiple environments in parallel
-func parallelLoadEnvironments(paths []string, opts ParallelOpts) (envs []*v1alpha1.Environment, err error) {
+func parallelLoadEnvironments(paths []string, opts parallelOpts) ([]*v1alpha1.Environment, error) {
 	wg := sync.WaitGroup{}
 	envsChan := make(chan parallelJob)
-	var allErrors []error
+	outChan := make(chan parallelOut)
 
-	numParallel := parallel
-	if opts.Parallel > 0 {
-		numParallel = opts.Parallel
+	if opts.Parallel <= 0 {
+		opts.Parallel = parallel
 	}
-	for i := 0; i < numParallel; i++ {
+
+	for i := 0; i < opts.Parallel; i++ {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			errs := parallelWorker(envsChan)
-			if errs != nil {
-				allErrors = append(allErrors, errs...)
-			}
+			parallelWorker(envsChan, outChan)
 		}()
 	}
 
-	results := make([]*v1alpha1.Environment, 0, len(paths))
-
+	jobs := 0
 	for _, path := range paths {
-		env := &v1alpha1.Environment{}
-		results = append(results, env)
 		envsChan <- parallelJob{
 			path: path,
-			opts: opts.JsonnetOpts,
-			env:  env,
+			opts: Opts{JsonnetOpts: opts.JsonnetOpts},
 		}
+		jobs++
 	}
 	close(envsChan)
-	wg.Wait()
 
-	for _, env := range results {
-		if env == nil {
+	var envs []*v1alpha1.Environment
+	var errors []error
+	for i := 0; i < jobs; i++ {
+		out := <-outChan
```

Comment:
You read exactly `len(jobs)` times from the channel. Your worker routines write in sum `len(jobs)` times to the channel, so once your loop is finished, the channel is fully drained and all worker routines exited